yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala

```diff
@@ -313,6 +320,21 @@ private[spark] class Client(
     env("SPARK_YARN_MODE") = "true"
     env("SPARK_YARN_STAGING_DIR") = stagingDir
     env("SPARK_USER") = UserGroupInformation.getCurrentUser().getShortUserName()
+    // If we logged in from keytab, make sure we copy the keytab to the staging directory on
+    // HDFS, and setup the relevant environment vars, so the AM can login again.
+    if (loginFromKeytab) {
+      val fs = FileSystem.get(hadoopConf)
+      val stagingDirPath = new Path(fs.getHomeDirectory, stagingDir)
+      val localUri = new URI(args.keytab)
+      val localPath = getQualifiedLocalPath(localUri, hadoopConf)
+      val destinationPath = new Path(stagingDirPath, keytabFileName)
+      val replication = sparkConf.getInt("spark.yarn.submit.file.replication",
```

Comment:
we already get this in spark.yarn.submit.file.replication, would be nice to only get it once and reuse.

Seems like this would belong in prepareLocalResources, just change description of it slightly.